addons/product/tests/test_product_attribute_value_config.py

```diff
@@ -712,3 +712,38 @@ def test_copy_extra_prices_of_product_attribute_values(self):
             'price_extra'
         )
         self.assertEqual(extra_prices, copied_extra_prices)
+
+    def test_inactive_product_with_dynamic_attribute_values(self):
+        product_attribut = self.env['product.attribute'].create({
+            'name': 'Color',
+            'sequence': 1,
+            'create_variant': 'dynamic',
+        })
+        attr_value = self.env['product.attribute.value'].create({
+            'name': 'Blue',
+            'attribute_id': product_attribut.id,
+            'sequence': 1,
+        })
+        first_product = self.env['product.template'].create({
+            'name': 'Sofa',
+            'detailed_type': 'consu',
+        })
+        first_product.write({
+            'attribute_line_ids': [(0, 0, {
+            'attribute_id': product_attribut.id,
+            'value_ids': [(6, 0, [attr_value.id])],
+            })]
+        })
+        second_product = first_product.copy({
+            'product_variant_ids': [(0, 0, {
+            'name': 'Sofa',
+            })]
+        })
+        products = self.env['product.template'].search([('id', 'in', (first_product.id, second_product.id))])
+        products.action_archive()
+        self.assertFalse(products[0].active, 'The product should be archived.')
+        self.assertFalse(products[1].active, 'The product should be archived.')
+        products.action_unarchive()
+        # check products should be unarchived successfully.
+        self.assertTrue(products[0].active, 'The product should be unarchived.')
+        self.assertTrue(products[1].active, 'The product should be unarchived.')
```

Comment:
Should move test cases to test_variants.py https://github.com/odoo/odoo/blob/a726966b8f61255ed39dc5b23422b5a5ac4699e4/addons/product/tests/test_variants.py#L356

Write a test case like below:
```
def test_unarchive_multiple_products_with_variants(self):
        product_attribut = self.env['product.attribute'].create({
            'name': 'Color',
            'sequence': 1,
            'create_variant': 'dynamic',
        })
        attr_value = self.env['product.attribute.value'].create({
            'name': 'Blue',
            'attribute_id': product_attribut.id,
            'sequence': 1,
        })
        first_product = self.env['product.template'].create({
            'name': 'Sofa',
            'attribute_line_ids': [(0, 0, {
                'attribute_id': product_attribut.id,
                'value_ids': [(6, 0, [attr_value.id])],
            })]
        })
        second_product = first_product.copy({
            'product_variant_ids': [(0, 0, {
                'name': 'Sofa variant 1',
            })]
        })

        products = first_product + second_product
        products.action_archive()
        self.assertFalse(first_product.active)
        self.assertFalse(second_product.active)
        self.assertFalse(second_product.product_variant_ids)

        products.action_unarchive()
        
        # check products should be unarchived successfully.
        self.assertTrue(first_product.active)
        self.assertTrue(second_product.active)
        self.assertTrue(second_product.product_variant_ids)
```